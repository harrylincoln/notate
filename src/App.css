a {
  color: #a0ae61;
}

.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
  pointer-events: none;
}

.App-content {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-content .canvas-container {
  display: flex
}

.App-content button {
  cursor: pointer
}

.App-content .canvas-container button:disabled {
  cursor: not-allowed;
}

.App-content .canvas-container canvas {
  background-color: white
}

.App-content .cof-container {
  width: 30%;
  margin: 0 auto;
}

.App-content .cof-container img {
  width: 100%
}

.App-content .toggle-controls-container {
  margin-top: 0.5rem;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  background-color: grey;
  padding: 0.5rem;
  color: black;
}

.App-content .toggle-controls-container label {
  margin-right: 1rem, 
}

.App-content .toggle-controls-container p {
  align-self: center;
  color: white;
  cursor: pointer;
  margin: 0;
}

.App-content .toggle-controls-container button {
  align-self: center;
  padding: 0.5rem;
  margin: 0.25rem;
}

.App-link {
  color: #61dafb;
}

.App-nav {
  position: fixed;
  top: 0;
  width: 100%;
  background: white;
}

.App-nav-list {
  margin: 0;
  padding: 0.25rem;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  list-style: none;
  width: 100%;
  border-bottom: 1px solid #282c34;
}

.App-nav-list li {
  margin: 0 1rem;
}

.flx {
  display: flex;
}

.code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
