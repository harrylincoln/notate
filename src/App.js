import './App.css';
import 'react-toastify/dist/ReactToastify.css';
import React from 'react';
import ReactGA from 'react-ga';
import { ToastContainer } from 'react-toastify';
import {
  retrieveUserData,
  writeToUserData
} from './utils/user-data';

import BarView from './components/bar-view';


class App extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      userData: null,
    };
    this.updateUserEnteredKey = this.updateUserEnteredKey.bind(this);
    this.updateUserNavigation = this.updateUserNavigation.bind(this);
    this.updateUserData = this.updateUserData.bind(this);

  }

  initialiseUserData() {
    return new Promise(resolve => this.setState({
      userData: retrieveUserData()
    }, resolve()));
  }

  async componentWillMount() {
    await this.initialiseUserData();
  }

  componentDidMount() {
    ReactGA.initialize('UA-146065324-1');
    ReactGA.pageview('/home');
  }

  updateUserEnteredKey(e) {
    let userData = {...this.state.userData};
    userData.userKey = e.target.value.toLowerCase();
    this.setState({userData});
  }

  updateUserNavigation(direction) {
    let userData = {...this.state.userData};
    direction === '+' ? userData.appStep += 1 : userData.appStep += -1;
    this.setState({userData});
  }

  updateUserData(data) {
    const userData = Object.assign(this.state.userData, data);
    this.setState({userData}, () => writeToUserData(userData));
  }

  render() {

    const { 
      userData
    } = this.state;
      return (
        <>
        <div className="App">
          <ToastContainer />
          <nav className="App-nav">
            <ul className="App-nav-list">
            <li>Active key: {userData.userKey.toUpperCase()}</li>
              {userData.activeBarNumber && (
                <li>Bar #: {userData.activeBarNumber}</li>
              )}
            </ul>
          </nav>
          {userData.appStep === 0 && (
            <section style={{padding: '2rem'}} className="App-content">
              <div className="cof-container">
                <img src="./circle-of-fifths.jpg" alt="circle of fifths"/>
              </div>
              <h1 style={{marginTop: '0'}}>Notes to tab</h1>
              <h4 style={{marginTop: '0'}}>Choose a key (uppercase only):</h4>
              <input onChange={this.updateUserEnteredKey} value={userData.userKey.toUpperCase()} placeholder='C by default' />
              <button style={{marginTop: '0.5rem'}} onClick={() => this.updateUserNavigation('+')}>Next Step</button>
              <h4>Usage / gotchas:</h4>

              <p style={{marginTop: '0'}}>- Use a laptop, not a phone. I mean, it might work but not tested <span role='img' aria-label='shrug'>🤷</span></p>

              <p>- Accidentals are not implied based on previously declared ones. 
              ie. if you want another flat outside of the key after already adding one 
              at the beginning of the bar, you must implicitly set it again.</p>

              <p>- 4/4 is the only time sig at the moment</p>

              <p>- It has a tough time if you plot tightly voiced triads/chords. This is a known bug and will be fixed.</p>

              <p>- Triplet grouping, beams, dotted values, hooks and tailing of notes aren't configurable at this time. Group them tight if you want a triplet <span role='img' aria-label='kiss'>😚</span></p>

              <p>- Tab has no concept of implicitly setting rests between notes. To address this, your grid will snap the notes to the beat lines.</p>

              <p>- If you'd like to raise a bug/PR please do so via the <a rel="noopener noreferrer" target="_blank" href="https://github.com/harrylincoln/notate/">project's Github repo</a> <span role='img' aria-label='thanks'>🙏</span></p>
            </section>
          )}
          {userData.appStep === 1 && (
            <BarView 
            userKey={userData.userKey} 
            updateUserData={this.updateUserData} 
            userData={userData} />
          )}
        </div>
        </>
      )
  }
}

export default App;
