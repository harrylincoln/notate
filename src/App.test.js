import React from 'react';
import ReactGA from 'react-ga';
import { shallow, mount } from 'enzyme';
import App from './App';

jest.mock("react-ga");
HTMLCanvasElement.prototype.getContext = () => jest.func();

const setStateSpy = jest.spyOn(App.prototype, 'setState');

afterEach(() => {    
  jest.clearAllMocks();
});

test('App snapshot rendering', () => {
  const wrapper = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
});

test('should initialise userData if no userData', () => {
  const getItemSpy = jest.spyOn(Storage.prototype, 'getItem');
  
  shallow(<App />);

  expect(getItemSpy).toHaveBeenCalled();
  expect(setStateSpy).toHaveBeenCalledWith({userData: {
    appStep: 0,
    userKey: 'C'
  }});
});

test('should initialise analytics', () => {
  const reactGaInitSpy = jest.spyOn(ReactGA, 'initialize');
  const reactGaPgViewSpy = jest.spyOn(ReactGA, 'pageview');

  shallow(<App />);
  expect(reactGaInitSpy).toHaveBeenCalledWith('UA-146065324-1');
  expect(reactGaPgViewSpy).toHaveBeenCalledWith('/home');

});

test('should update user key when continuing through to bar-view', () => {
  const newKey = 'D';
  const wrapper = mount(<App />);
  wrapper.find('#userKeyInput').simulate('change', { target: { value: newKey } });
  expect(setStateSpy).toHaveBeenCalledWith({userData: {
    appStep: 0,
    userKey: newKey
  }});
  wrapper.find('#continueToBarViewBtn').simulate('click');
  expect(setStateSpy).toHaveBeenCalledWith({userData: {
    activeBarNumber: 1,
    appStep: 1,
    userKey: newKey
  }});
});