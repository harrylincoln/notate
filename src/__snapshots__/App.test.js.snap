// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot rendering 1`] = `ShallowWrapper {}`;
